promql/engine.go

```diff
@@ -943,6 +954,7 @@ func (ev *evaluator) rangeEval(f func([]parser.Value, *EvalNodeHelper) Vector, e
 		enh.out = result[:0] // Reuse result vector.
 
 		ev.currentSamples += len(result)
+		ev.samplesStats.IncrementPeak(len(result))
```

Comment:
You're assuming that we're at the peak right now, which is not a good assumption. You need to check if currentSamples is actually the peak now, and if so update.

You should only ever be setting the peak, not incrementing it.

Reply:
so, does that mean that peak is just the max number samples added to the result at an instant among many instances (and just catch those numbers and not increment as incrementing will be mostly same as total samples)?

Reply:
No, it's the peak value of currentSamples. Consider for example if a big aggregation set the peak to 1M, and we've now aggregated down to a single sample. No further processing will increase the samples in memory beyond the peak.

Reply:
Updated. How does it look now?

Reply:
That's not it, the only number that should ever be considered for updating the peak is currentSamples as it's the peak value of currentSamples that we're trying to capture.

Reply:
That's what is being considered, or the increase of it. Any other places you see please comment.

Reply:
You're not passing in currentSamples here, you're passing in something else. The biggest increase to currentSamples isn't interesting, it's the highest value it has that is interesting.